src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -2,89 +2,147 @@
 
 import com.epam.aidial.core.ProxyContext;
 import com.epam.aidial.core.config.ApiKeyData;
-import com.google.common.annotations.VisibleForTesting;
+import com.epam.aidial.core.data.ResourceType;
+import com.epam.aidial.core.service.ResourceService;
+import com.epam.aidial.core.storage.ResourceDescription;
+import com.epam.aidial.core.util.EtagHeader;
+import com.epam.aidial.core.util.ProxyUtil;
 import io.vertx.core.Future;
-import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
 
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 
+import static com.epam.aidial.core.storage.BlobStorageUtil.PATH_SEPARATOR;
+
+@Slf4j
+@RequiredArgsConstructor
 public class TokenStatsTracker {
```

Comment:
side track: seems like current algo can leave junk on storage

Reply:
right. that may happen if root span is not completed for some reason: pod restart or something else